File: 0140-word-break-ii/0140-word-break-ii.cpp
class Solution {
public:
    vector<string> result;
    unordered_set<string> dict;

    void solve(int i, string &curr, string &s) {
        if (i >= s.length()) {
            result.push_back(curr);
            return;
        }
        for (int j = i; j < s.length(); j++) {
            string tempWord = s.substr(i, j - i + 1);
            if (dict.find(tempWord) != dict.end()) {
                string origSentence = curr;
                if (!curr.empty()) {
                    curr += " ";
                }
                curr += tempWord;
                solve(j+1, curr, s);
                curr = origSentence;
            }
        }
    }

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        for (string &word : wordDict) {
            dict.insert(word);
        }
        string curr = "";
        solve(0, curr, s);
        return result;
    }
};

File: LeetCode/0140-word-break-ii/0140-word-break-ii.java
class Solution {
    private List<String> result = new ArrayList<>();
    private Set<String> dict = new HashSet<>();

    public void solve(int i, String curr, String s) {
        if (i >= s.length()) {
            result.add(curr.trim());
            return;
        }
        for (int j = i; j < s.length(); j++) {
            String temp = s.substring(i, j + 1);
            if (dict.contains(temp)) {
                String original = curr;
                if (!curr.isEmpty()) {
                    curr += " ";
                }
                curr += temp;
                solve(j + 1, curr, s);
                curr = original;
            }
        }
    }

    public List<String> wordBreak(String s, List<String> wordDict) {
        dict.addAll(wordDict);
        solve(0, "", s);
        return result;
    }
}